domain/schema_validator.go

```diff
@@ -206,3 +191,32 @@ func extractPhysicalTime(ts uint64) time.Time {
 	t := int64(ts >> 18) // 18 for physicalShiftBits
 	return time.Unix(t/1e3, (t%1e3)*1e6)
 }
+
+// simpleQueue is a fixed size queue, when queue is full, enqueue will overwrite the oldest.
+// empty, initial state: head == tail
+// queue full:  (tail+1) % size == head
+type simpleQueue struct {
+	data [maxNumberOfDiffsToLoad]deltaSchemaInfo
+	head int
+	tail int
+}
+
+func (q *simpleQueue) enqueue(schemaVersion int64, relatedTableIDs []int64) {
+	// Move tail cursor forward.
+	q.tail = (q.tail + 1) % maxNumberOfDiffsToLoad
```

Comment:
since you store the oldest value in `q.data[q.head]`, see line 130 in domain/schema_validator.go, that indicates `q.data[q.head]` stores element and `q.data[q.tail]` should not store element

you should store element in `q.data[q.tail]` first then increase `q.tail` by 1

when you get tail value, see line 151 in domain/schema_validator_test.go, you should return `q.data[q.tail-1]`